// starknet/provider/default.js
'use strict';
var __assign =
  (this && this.__assign) ||
  function () {
    __assign =
      Object.assign ||
      function (t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
          s = arguments[i];
          for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
      };
    return __assign.apply(this, arguments);
  };
var __awaiter =
  (this && this.__awaiter) ||
  function (thisArg, _arguments, P, generator) {
    function adopt(value) {
      return value instanceof P
        ? value
        : new P(function (resolve) {
            resolve(value);
          });
    }
    return new (P || (P = Promise))(function (resolve, reject) {
      function fulfilled(value) {
        try {
          step(generator.next(value));
        } catch (e) {
          reject(e);
        }
      }
      function rejected(value) {
        try {
          step(generator['throw'](value));
        } catch (e) {
          reject(e);
        }
      }
      function step(result) {
        result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
      }
      step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
  };
var __generator =
  (this && this.__generator) ||
  function (thisArg, body) {
    var _ = {
        label: 0,
        sent: function () {
          if (t[0] & 1) throw t[1];
          return t[1];
        },
        trys: [],
        ops: [],
      },
      f,
      y,
      t,
      g;
    return (
      (g = { next: verb(0), throw: verb(1), return: verb(2) }),
      typeof Symbol === 'function' &&
        (g[Symbol.iterator] = function () {
          return this;
        }),
      g
    );
    function verb(n) {
      return function (v) {
        return step([n, v]);
      };
    }
    function step(op) {
      if (f) throw new TypeError('Generator is already executing.');
      while (_)
        try {
          if (
            ((f = 1),
            y &&
              (t =
                op[0] & 2
                  ? y['return']
                  : op[0]
                  ? y['throw'] || ((t = y['return']) && t.call(y), 0)
                  : y.next) &&
              !(t = t.call(y, op[1])).done)
          )
            return t;
          if (((y = 0), t)) op = [op[0] & 2, t.value];
          switch (op[0]) {
            case 0:
            case 1:
              t = op;
              break;
            case 4:
              _.label++;
              return { value: op[1], done: false };
            case 5:
              _.label++;
              y = op[1];
              op = [0];
              continue;
            case 7:
              op = _.ops.pop();
              _.trys.pop();
              continue;
            default:
              if (
                !((t = _.trys), (t = t.length > 0 && t[t.length - 1])) &&
                (op[0] === 6 || op[0] === 2)
              ) {
                _ = 0;
                continue;
              }
              if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                _.label = op[1];
                break;
              }
              if (op[0] === 6 && _.label < t[1]) {
                _.label = t[1];
                t = op;
                break;
              }
              if (t && _.label < t[2]) {
                _.label = t[2];
                _.ops.push(op);
                break;
              }
              if (t[2]) _.ops.pop();
              _.trys.pop();
              continue;
          }
          op = body.call(thisArg, _);
        } catch (e) {
          op = [6, e];
          y = 0;
        } finally {
          f = t = 0;
        }
      if (op[0] & 5) throw op[1];
      return { value: op[0] ? op[1] : void 0, done: true };
    }
  };
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, '__esModule', { value: true });
exports.Provider = void 0;
var axios_1 = __importDefault(require('axios'));
var url_join_1 = __importDefault(require('url-join'));
var json_1 = require('../utils/json');
var number_1 = require('../utils/number');
var stark_1 = require('../utils/stark');
var utils_1 = require('./utils');
function wait(delay) {
  return new Promise(function (res) {
    return setTimeout(res, delay);
  });
}
var Provider = /** @class */ (function () {
  function Provider(optionsOrProvider) {
    if (optionsOrProvider === void 0) {
      optionsOrProvider = { network: 'goerli-alpha' };
    }
    if (optionsOrProvider instanceof Provider) {
      this.baseUrl = optionsOrProvider.baseUrl;
      this.feederGatewayUrl = optionsOrProvider.feederGatewayUrl;
      this.gatewayUrl = optionsOrProvider.gatewayUrl;
    } else {
      var baseUrl =
        'baseUrl' in optionsOrProvider
          ? optionsOrProvider.baseUrl
          : Provider.getNetworkFromName(optionsOrProvider.network);
      this.baseUrl = baseUrl;
      this.feederGatewayUrl = (0, url_join_1.default)(baseUrl, 'feeder_gateway');
      this.gatewayUrl = (0, url_join_1.default)(baseUrl, 'gateway');
    }
  }
  Provider.getNetworkFromName = function (name) {
    switch (name) {
      case 'mainnet-alpha':
        return 'https://alpha-mainnet.starknet.io';
      case 'goerli-alpha':
      default:
        return 'https://alpha4.starknet.io';
    }
  };
  /**
   * Gets the smart contract address on the goerli testnet.
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/f464ec4797361b6be8989e36e02ec690e74ef285/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L13-L15)
   * @returns starknet smart contract addresses
   */
  Provider.prototype.getContractAddresses = function () {
    return __awaiter(this, void 0, void 0, function () {
      var data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            return [
              4 /*yield*/,
              axios_1.default.get(
                (0, url_join_1.default)(this.feederGatewayUrl, 'get_contract_addresses')
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  /**
   * Calls a function on the StarkNet contract.
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/fc97bdd8322a7df043c87c371634b26c15ed6cee/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L25-L39)
   *
   * @param invokeTransaction - transaction to be invoked
   * @param blockHash
   * @param blockNumber
   * @returns the result of the function on the smart contract.
   */
  Provider.prototype.callContract = function (invokeTransaction, blockIdentifier) {
    if (blockIdentifier === void 0) {
      blockIdentifier = null;
    }
    return __awaiter(this, void 0, void 0, function () {
      var formattedBlockIdentifier, data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            formattedBlockIdentifier = (0, utils_1.getFormattedBlockIdentifier)(blockIdentifier);
            return [
              4 /*yield*/,
              axios_1.default.post(
                (0, url_join_1.default)(
                  this.feederGatewayUrl,
                  'call_contract',
                  formattedBlockIdentifier
                ),
                __assign({ signature: [], calldata: [] }, invokeTransaction)
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  /**
   * Gets the block information
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/fc97bdd8322a7df043c87c371634b26c15ed6cee/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L41-L53)
   *
   * @param blockHash
   * @param blockNumber
   * @returns the block object { block_number, previous_block_number, state_root, status, timestamp, transaction_receipts, transactions }
   */
  Provider.prototype.getBlock = function (blockIdentifier) {
    if (blockIdentifier === void 0) {
      blockIdentifier = null;
    }
    return __awaiter(this, void 0, void 0, function () {
      var formattedBlockIdentifier, data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            formattedBlockIdentifier = (0, utils_1.getFormattedBlockIdentifier)(blockIdentifier);
            return [
              4 /*yield*/,
              axios_1.default.get(
                (0, url_join_1.default)(
                  this.feederGatewayUrl,
                  'get_block',
                  formattedBlockIdentifier
                )
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  /**
   * Gets the code of the deployed contract.
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/fc97bdd8322a7df043c87c371634b26c15ed6cee/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L55-L68)
   *
   * @param contractAddress
   * @param blockHash
   * @param blockNumber
   * @returns Bytecode and ABI of compiled contract
   */
  Provider.prototype.getCode = function (contractAddress, blockIdentifier) {
    if (blockIdentifier === void 0) {
      blockIdentifier = null;
    }
    return __awaiter(this, void 0, void 0, function () {
      var formattedBlockIdentifier, data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            formattedBlockIdentifier = (0, utils_1.getFormattedBlockIdentifier)(blockIdentifier);
            return [
              4 /*yield*/,
              axios_1.default.get(
                (0, url_join_1.default)(
                  this.feederGatewayUrl,
                  'get_code',
                  '?contractAddress=' + contractAddress + '&' + formattedBlockIdentifier
                )
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  // TODO: add proper type
  /**
   * Gets the contract's storage variable at a specific key.
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/fc97bdd8322a7df043c87c371634b26c15ed6cee/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L70-L85)
   *
   * @param contractAddress
   * @param key - from getStorageVarAddress('<STORAGE_VARIABLE_NAME>') (WIP)
   * @param blockHash
   * @param blockNumber
   * @returns the value of the storage variable
   */
  Provider.prototype.getStorageAt = function (contractAddress, key, blockIdentifier) {
    if (blockIdentifier === void 0) {
      blockIdentifier = null;
    }
    return __awaiter(this, void 0, void 0, function () {
      var formattedBlockIdentifier, data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            formattedBlockIdentifier = (0, utils_1.getFormattedBlockIdentifier)(blockIdentifier);
            return [
              4 /*yield*/,
              axios_1.default.get(
                (0, url_join_1.default)(
                  this.feederGatewayUrl,
                  'get_storage_at',
                  '?contractAddress=' +
                    contractAddress +
                    '&key=' +
                    key +
                    '&' +
                    formattedBlockIdentifier
                )
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  /**
   * Gets the status of a transaction.
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/f464ec4797361b6be8989e36e02ec690e74ef285/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L48-L52)
   *
   * @param txHash
   * @returns the transaction status object { block_number, tx_status: NOT_RECEIVED | RECEIVED | PENDING | REJECTED | ACCEPTED_ONCHAIN }
   */
  Provider.prototype.getTransactionStatus = function (txHash) {
    return __awaiter(this, void 0, void 0, function () {
      var txHashBn, data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            txHashBn = (0, number_1.toBN)(txHash);
            return [
              4 /*yield*/,
              axios_1.default.get(
                (0, url_join_1.default)(
                  this.feederGatewayUrl,
                  'get_transaction_status',
                  '?transactionHash=' + (0, number_1.toHex)(txHashBn)
                )
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  /**
   * Gets the transaction information from a tx id.
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/f464ec4797361b6be8989e36e02ec690e74ef285/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L54-L58)
   *
   * @param txHash
   * @returns the transacton object { transaction_id, status, transaction, block_number?, block_number?, transaction_index?, transaction_failure_reason? }
   */
  Provider.prototype.getTransaction = function (txHash) {
    return __awaiter(this, void 0, void 0, function () {
      var txHashBn, data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            txHashBn = (0, number_1.toBN)(txHash);
            return [
              4 /*yield*/,
              axios_1.default.get(
                (0, url_join_1.default)(
                  this.feederGatewayUrl,
                  'get_transaction',
                  '?transactionHash=' + (0, number_1.toHex)(txHashBn)
                )
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  /**
   * Invoke a function on the starknet contract
   *
   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/f464ec4797361b6be8989e36e02ec690e74ef285/src/starkware/starknet/services/api/gateway/gateway_client.py#L13-L17)
   *
   * @param transaction - transaction to be invoked
   * @returns a confirmation of invoking a function on the starknet contract
   */
  Provider.prototype.addTransaction = function (transaction) {
    return __awaiter(this, void 0, void 0, function () {
      var signature, contract_address_salt, data;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            signature =
              transaction.type === 'INVOKE_FUNCTION' &&
              (0, stark_1.formatSignature)(transaction.signature);
            contract_address_salt =
              transaction.type === 'DEPLOY' &&
              (0, number_1.toHex)((0, number_1.toBN)(transaction.contract_address_salt));
            return [
              4 /*yield*/,
              axios_1.default.post(
                (0, url_join_1.default)(this.gatewayUrl, 'add_transaction'),
                (0, json_1.stringify)(
                  __assign(
                    __assign(
                      __assign({}, transaction),
                      Array.isArray(signature) && { signature: signature }
                    ),
                    contract_address_salt && { contract_address_salt: contract_address_salt }
                  )
                ),
                { headers: { 'Content-Type': 'application/json' } }
              ),
            ];
          case 1:
            data = _a.sent().data;
            return [2 /*return*/, data];
        }
      });
    });
  };
  /**
   * Deploys a given compiled contract (json) to starknet
   *
   * @param contract - a json object containing the compiled contract
   * @param address - (optional, defaults to a random address) the address where the contract should be deployed (alpha)
   * @returns a confirmation of sending a transaction on the starknet contract
   */
  Provider.prototype.deployContract = function (contract, constructorCalldata, addressSalt) {
    if (constructorCalldata === void 0) {
      constructorCalldata = [];
    }
    if (addressSalt === void 0) {
      addressSalt = (0, stark_1.randomAddress)();
    }
    var parsedContract = typeof contract === 'string' ? (0, json_1.parse)(contract) : contract;
    var contractDefinition = __assign(__assign({}, parsedContract), {
      program: (0, stark_1.compressProgram)(parsedContract.program),
    });
    return this.addTransaction({
      type: 'DEPLOY',
      contract_address_salt: addressSalt,
      constructor_calldata: constructorCalldata,
      contract_definition: contractDefinition,
    });
  };
  /**
   * Invokes a function on starknet
   *
   * @param contractAddress - target contract address for invoke
   * @param entrypointSelector - target entrypoint selector for
   * @param calldata - (optional, default []) calldata
   * @param signature - (optional) signature to send along
   * @returns response from addTransaction
   */
  Provider.prototype.invokeFunction = function (
    contractAddress,
    entrypointSelector,
    calldata,
    signature
  ) {
    return this.addTransaction({
      type: 'INVOKE_FUNCTION',
      contract_address: contractAddress,
      entry_point_selector: entrypointSelector,
      calldata: calldata,
      signature: signature,
    });
  };
  Provider.prototype.waitForTx = function (txHash, retryInterval) {
    if (retryInterval === void 0) {
      retryInterval = 8000;
    }
    return __awaiter(this, void 0, void 0, function () {
      var onchain, res, error;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            onchain = false;
            return [4 /*yield*/, wait(retryInterval)];
          case 1:
            _a.sent();
            _a.label = 2;
          case 2:
            if (!!onchain) return [3 /*break*/, 5];
            // eslint-disable-next-line no-await-in-loop
            return [4 /*yield*/, wait(retryInterval)];
          case 3:
            // eslint-disable-next-line no-await-in-loop
            _a.sent();
            return [4 /*yield*/, this.getTransactionStatus(txHash)];
          case 4:
            res = _a.sent();
            if (res.tx_status === 'ACCEPTED_ON_L1' || res.tx_status === 'ACCEPTED_ON_L2') {
              onchain = true;
            } else if (res.tx_status === 'REJECTED' || res.tx_status === 'NOT_RECEIVED') {
              error = Error(res.tx_status);
              error.response = res;
              throw error;
            }
            return [3 /*break*/, 2];
          case 5:
            return [2 /*return*/];
        }
      });
    });
  };
  return Provider;
})();
exports.Provider = Provider;
