// @shardlabs/starknet-hardhat-plugin/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PENDING_BLOCK_NUMBER = exports.SHORT_STRING_MAX_CHARACTERS = exports.VOYAGER_MAINNET_CONTRACT_API_URL = exports.VOYAGER_GOERLI_CONTRACT_API_URL = exports.LEN_SUFFIX = exports.CHECK_STATUS_RECOVER_TIMEOUT = exports.CHECK_STATUS_TIMEOUT = exports.ALPHA_MAINNET_URL = exports.ALPHA_URL = exports.DEFAULT_STARKNET_NETWORK = exports.ALPHA_MAINNET_INTERNALLY = exports.ALPHA_MAINNET = exports.ALPHA_TESTNET_INTERNALLY = exports.ALPHA_TESTNET = exports.GITHUB_ACCOUNT_ARTIFACTS_URL = exports.ACCOUNT_ARTIFACTS_VERSION = exports.ACCOUNT_CONTRACT_ARTIFACTS_ROOT_PATH = exports.DEFAULT_DOCKER_IMAGE_TAG = exports.DOCKER_REPOSITORY = exports.DEFAULT_STARKNET_ACCOUNT_PATH = exports.DEFAULT_STARKNET_ARTIFACTS_PATH = exports.DEFAULT_STARKNET_SOURCES_PATH = exports.ABI_SUFFIX = exports.PLUGIN_NAME = void 0;
exports.PLUGIN_NAME = "Starknet";
exports.ABI_SUFFIX = "_abi.json";
exports.DEFAULT_STARKNET_SOURCES_PATH = "contracts";
exports.DEFAULT_STARKNET_ARTIFACTS_PATH = "starknet-artifacts";
exports.DEFAULT_STARKNET_ACCOUNT_PATH = "~/.starknet_accounts";
exports.DOCKER_REPOSITORY = "shardlabs/cairo-cli";
exports.DEFAULT_DOCKER_IMAGE_TAG = "0.7.1";
exports.ACCOUNT_CONTRACT_ARTIFACTS_ROOT_PATH = "account-contract-artifacts";
exports.ACCOUNT_ARTIFACTS_VERSION = "0.1.0";
exports.GITHUB_ACCOUNT_ARTIFACTS_URL = `https://raw.githubusercontent.com/Shard-Labs/starknet-hardhat-example/plugin/${exports.ACCOUNT_CONTRACT_ARTIFACTS_ROOT_PATH}/${exports.ACCOUNT_ARTIFACTS_VERSION}/`;
exports.ALPHA_TESTNET = "alpha-goerli";
exports.ALPHA_TESTNET_INTERNALLY = "alpha";
exports.ALPHA_MAINNET = "alpha-mainnet";
exports.ALPHA_MAINNET_INTERNALLY = "alphaMainnet";
exports.DEFAULT_STARKNET_NETWORK = exports.ALPHA_TESTNET_INTERNALLY;
exports.ALPHA_URL = "https://alpha4.starknet.io";
exports.ALPHA_MAINNET_URL = "https://alpha-mainnet.starknet.io";
exports.CHECK_STATUS_TIMEOUT = 5000; // ms
exports.CHECK_STATUS_RECOVER_TIMEOUT = 10000; // ms
exports.LEN_SUFFIX = "_len";
exports.VOYAGER_GOERLI_CONTRACT_API_URL = "https://goerli.voyager.online/api/contract/";
exports.VOYAGER_MAINNET_CONTRACT_API_URL = "https://voyager.online/api/contract/";
exports.SHORT_STRING_MAX_CHARACTERS = 31;
exports.PENDING_BLOCK_NUMBER = "pending";
